File: hulkify/hulkify.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# system imports
import sys
import pkg_resources
import re
import random
import json

# library imports
import pattern.en

try:
    data = pkg_resources.resource_string(__name__, "./grammar.json")
    GRAMMAR = json.loads(data)
except:
    raise RuntimeError("Couldn't load hulkify grammar file.")


# preprocess to account for unicode
def fixList(myList):
    replacement = []
    for entry in myList:
        replacement.append(entry)
        if u"'" in entry:
            replacement.append(entry.replace(u"'", u"’"))
    return replacement

def fixDict(myDict):
    replacement = {}
    for key, value in myDict.iteritems():
        replacement[key] = value
        if u"'" in key:
            replacement[key.replace(u"'", u"’")] = value
    return replacement

GRAMMAR["PERSONAL_PRONOUNS"] = fixList(GRAMMAR["PERSONAL_PRONOUNS"])
GRAMMAR["PERSONAL_POSSESSIVE_PRONOUNS"] = fixList(GRAMMAR["PERSONAL_POSSESSIVE_PRONOUNS"])
GRAMMAR["CONTRACTION_MAPPING"] = fixDict(GRAMMAR["CONTRACTION_MAPPING"])
GRAMMAR["LINKING_VERBS"] = fixList(GRAMMAR["LINKING_VERBS"])
GRAMMAR["ARTICLES"] = fixList(GRAMMAR["ARTICLES"])
GRAMMAR["DIMINUTIVE_ADJECTIVES"] = fixList(GRAMMAR["DIMINUTIVE_ADJECTIVES"])


def hulkify(bannerText, maxLength=None, encoding="utf-8"):
    if type(bannerText) != unicode:
        hulkText = unicode(bannerText, encoding)
    else:
        hulkText = bannerText

    def removeWordFromMatch(matchObject):
        before = len(matchObject.group(1)) > 0
        after  = len(matchObject.group(3)) > 0

        if not before and not after:
            return matchObject.expand(ur"\2")
        if before and after:
            return matchObject.expand(ur"\1")
        if before:
            return matchObject.expand(ur"\3")
        if after:
            return matchObject.expand(ur"\1")

    def changeTextUnlessOverrun(oldText, newText, maximum):
        if maximum == None:
            return newText
        if len(newText) > maximum:
            return oldText
        return newText

    # Change contractions to their main word. ("can't" -> "not")
    for con, main in GRAMMAR["CONTRACTION_MAPPING"].iteritems():
        hulkText = re.sub(ur"(\s?|^)(\b%s\b)(\s?|$)" % (con), ur"\1%s\3" % main, hulkText, flags=re.IGNORECASE | re.UNICODE)

    # Drop all linking verbs ("am" or "to be")
    for lv in GRAMMAR["LINKING_VERBS"]:
        hulkText = re.sub(ur"(\s?|^)(\b%s\b)(\s?|$)" % (lv), removeWordFromMatch, hulkText, flags=re.IGNORECASE | re.UNICODE)

    # Remove articles.
    for a in GRAMMAR["ARTICLES"]:
        hulkText = re.sub(ur"(\s?|^)(\b%s\b)(\s?|$)" % (a), removeWordFromMatch, hulkText, flags=re.IGNORECASE | re.UNICODE)


    ### everything above here only removes content; after this we have to consider length

    # Replace all personal pronouns (I/me/I've/I'm/etc) with "Hulk"
    for pp in GRAMMAR["PERSONAL_PRONOUNS"]:
        hulkText = changeTextUnlessOverrun(
            hulkText,
            re.sub(ur"(\s?|^)(\b%s\b)(\s?|$)" % (pp), ur"\1Hulk\3", hulkText, flags=re.IGNORECASE | re.UNICODE),
            maxLength
        )
    for ppp in GRAMMAR["PERSONAL_POSSESSIVE_PRONOUNS"]:
        if u"’" in bannerText:
            hulkText = changeTextUnlessOverrun(
                hulkText,
                re.sub(ur"(\s?|^)(\b%s\b)(\s?|$)" % (ppp), ur"\1Hulk’s\3", hulkText, flags=re.IGNORECASE | re.UNICODE),
                maxLength
            )
        else:
            hulkText = changeTextUnlessOverrun(
                hulkText,
                re.sub(ur"(\s?|^)(\b%s\b)(\s?|$)" % (ppp), ur"\1Hulk's\3", hulkText, flags=re.IGNORECASE | re.UNICODE),
                maxLength
            )

    parsed = pattern.en.tag(hulkText, tokenize=False)

    # Change verbs to present tense.
    for wordData in parsed:
        if wordData[1][0] == u"V":
            hulkText = changeTextUnlessOverrun(
                hulkText,
                re.sub(ur"\b%s\b" % wordData[0], pattern.en.lemma(wordData[0]), hulkText, flags=re.IGNORECASE | re.UNICODE),
                maxLength
            )

    # Replace any diminuitive adjective with "puny."
    for wordData in parsed:
        if wordData[1][:2] == u"JJ":
            if wordData[0] in GRAMMAR["DIMINUTIVE_ADJECTIVES"]:
                hulkText = changeTextUnlessOverrun(
                    hulkText,
                    re.sub(ur"\b%s\b" % wordData[0], u"puny", hulkText, flags=re.IGNORECASE | re.UNICODE),
                    maxLength
                )

    # Exclamation points.
    def exclaim(matchObject):
        punct = matchObject.group(0)
        if u"." in punct and punct.count(u"!") == 0:
            # change out periods most of the time
            if (random.random() < 0.7):
                punct = u"!"

        # if we're already at the max, don't intensify
        if (maxLength != None) and (len(hulkText) >= maxLength):
            return punct

        # 30% chance of intensifying
        if (u"." not in punct) and (random.random() < 0.3):
            punct += u"!"
            # slight chance of double intensifying
            if (random.random() < 0.45):
                punct += u"!"
                if u"?" in punct:
                    punct += u"?!"

        return punct

    hulkText = re.sub(ur"([.?!]+)", exclaim, hulkText)

    # Easter egg at mention of "strong"
    strongest = u" Hulk is the strongest there is!"
    if re.search(ur"\bstrong\b", hulkText, re.IGNORECASE | re.UNICODE) != None:
        if (maxLength == None) or (len(hulkText) + len(strongest) <= maxLength):
            hulkText += strongest

    # Convert the whole thing to uppercase. 
    hulkText = hulkText.upper()
    
    return hulkText


if __name__ == '__main__':
    try:
        with open("../test_corpus.txt", "r") as corpusFile:
            corpus = corpusFile.read().split("\n")
    except:
        corpus = []

    for text in corpus:
        print hulkify(unicode(text, "utf-8"), maxLength=140).encode("utf-8")

File: hulkify/hulkify.js

var fs  = require('fs');
var pos = require('pos');
var nlp = require('nlp_compromise');
var ent = require('html-entities').AllHtmlEntities;

var GRAMMAR = require('./grammar.json');

module.exports = hulkify;

function fixList(list) {
  var replacement = Array();
  for (var i = 0; i < list.length; i++) {
    replacement.push(list[i]);
    if (list[i].indexOf("'") >= 0) {
      replacement.push(list[i].replace("'", "’"));
    }
  }
  return replacement;
}
function fixDict(dict) {
  var replacement = {};
  for (var key in dict) {
    replacement[key] = dict[key];
    if (key.indexOf("'") >= 0) {
      replacement[key.replace("'", "’")] = dict[key];
    }
  }
  return replacement;
}

GRAMMAR.PERSONAL_PRONOUNS = fixList(GRAMMAR.PERSONAL_PRONOUNS);
GRAMMAR.PERSONAL_POSSESSIVE_PRONOUNS = fixList(GRAMMAR.PERSONAL_POSSESSIVE_PRONOUNS);
GRAMMAR.CONTRACTION_MAPPING = fixDict(GRAMMAR.CONTRACTION_MAPPING);
GRAMMAR.LINKING_VERBS = fixList(GRAMMAR.LINKING_VERBS);
GRAMMAR.ARTICLES = fixList(GRAMMAR.ARTICLES);
GRAMMAR.DIMINUTIVE_ADJECTIVES = fixList(GRAMMAR.DIMINUTIVE_ADJECTIVES);

function hulkify(bannerText, maxLength) {
  var entProc = new ent();
  bannerText = entProc.decode(bannerText);
  
  var hulkText = bannerText;

  function removeWordFromMatch(match, g1, g2, g3) {
    var before = g1.length > 0;
    var after  = g3.length > 0;

    if (!before && !after) {
      return g2;
    }
    if (before && after) {
      return g1;
    }
    if (before) {
      return g3;
    }
    if (after) {
      return g1;
    }
  }

  function changeTextUnlessOverrun(oldText, newText, maximum) {
    if (maximum === undefined) {
      return newText;
    }
    if (newText.length > maximum) {
      return oldText;
    }
    return newText;
  }

  // Change contractions to their main word. ("can't" -> "not")
  for (var con in GRAMMAR.CONTRACTION_MAPPING) {
    var main = GRAMMAR.CONTRACTION_MAPPING[con];
    var re = new RegExp("(\\s?|^)(\\b" + con + "\\b)(\\s?|$)", "gi");
    hulkText = hulkText.replace(re, "$1" + main + "$3");
  }

  // Drop all linking verbs ("am" or "to be")
  for (var i=0; i < GRAMMAR.LINKING_VERBS.length; i++) {
    var lv = GRAMMAR.LINKING_VERBS[i];
    var re = new RegExp("(\\s?|^)(\\b" + lv + "\\b)(\\s?|$)", "gi");
    hulkText = hulkText.replace(re, removeWordFromMatch);
  }

  // Remove articles.
  for (var i=0; i < GRAMMAR.ARTICLES.length; i++) {
    var a = GRAMMAR.ARTICLES[i];
    var re = new RegExp("(\\s?|^)(\\b" + a + "\\b)(\\s?|$)", "gi");
    hulkText = hulkText.replace(re, removeWordFromMatch);
  }

  // // everything above here only removes content; after this we have to consider length

  // Replace all personal pronouns (I/me/I've/I'm/etc) with "Hulk"
  for (var i = 0; i < GRAMMAR.PERSONAL_PRONOUNS.length; i++) {
    var pp = GRAMMAR.PERSONAL_PRONOUNS[i];
    var re = new RegExp("(\\s?|^)(\\b" + pp + "\\b)(\\s?|$)", "gi");
    hulkText = changeTextUnlessOverrun(
        hulkText,
        hulkText.replace(re, "$1Hulk$3"), 
        maxLength
      );
  }
  for (var i = 0; i < GRAMMAR.PERSONAL_POSSESSIVE_PRONOUNS.length; i++) {
    var ppp = GRAMMAR.PERSONAL_POSSESSIVE_PRONOUNS[i];
    var re = new RegExp("(\\s?|^)(\\b" + ppp + "\\b)(\\s?|$)", "gi");
    if (ppp.indexOf("’") >= 0) {
      hulkText = changeTextUnlessOverrun(
          hulkText,
          hulkText.replace(re, "$1Hulk’s$3"), 
          maxLength
        );
    }
    else {
      hulkText = changeTextUnlessOverrun(
          hulkText,
          hulkText.replace(re, "$1Hulk's$3"), 
          maxLength
        );
    }
  }

  var words = new pos.Lexer().lex(hulkText);
  var parsed = new pos.Tagger().tag(words);

  for (var i = 0; i < parsed.length; i++) {
    var wordData = parsed[i];
    if (wordData[1][0] == 'V' && wordData[1] != 'VBG') {
      var re = new RegExp("\\b" + wordData[0] + "\\b", "gi");
      hulkText = changeTextUnlessOverrun(
        hulkText,
        hulkText.replace(re, nlp.verb(wordData[0]).conjugate().infinitive),
        maxLength
      );
    }
  }

  // Replace any diminutive adjective with "puny."
  for (var i = 0; i < parsed.length; i++) {
    var wordData = parsed[i];
    if (wordData[1].substring(0,2) == 'JJ') {
      if (GRAMMAR.DIMINUTIVE_ADJECTIVES.indexOf(wordData[0]) >= 0) {
        var re = new RegExp("\\b" + wordData[0] + "\\b", "gi");
        hulkText = changeTextUnlessOverrun(
          hulkText,
          hulkText.replace(re, "puny"),
          maxLength
        );
      }
    }
  }

  // Exclamation points
  function exclaim (match, g0, g1) {
    var punct = g0;
    if ( (punct.indexOf('.') >= 0) && ((punct.match(/!/g) || []).length == 0) ) {
      // change out periods most of the time
      if (Math.random() < 0.7) {
        punct = '!';
      }
    }

    // if we're already at the max, don't intensify
    if ( (maxLength !== undefined) && (hulkText.length >= maxLength) ) {
      return punct;
    }

    // 30% chance of intensifying
    if ( (punct.indexOf('.') < 0) && (Math.random() < 0.3) ) {
      punct += '!';
      // slight chance of double intensifying
      if (Math.random() < 0.45) {
        punct += '!';
        if (punct.indexOf('?') >= 0) {
          punct += '?!';
        }
      }
    }

    return punct + g1;
  }

  hulkText = hulkText.replace(/([.?!]+)(\s|$)/gi, exclaim);

  // Easter egg at mention of "strong"
  var strongest = ' Hulk is the strongest there is!';
  var strongRe = new RegExp('\\bstrong\\b', 'gi');
  if (strongRe.test(hulkText)) {
    if ( (maxLength === undefined) || (hulkText.length + strongest.length <= maxLength) ) {
      hulkText += strongest;
    }
  }

  // Convert the whole thing to uppercase.
  hulkText = hulkText.toUpperCase();

  return hulkText;
}


var main = function(){
  var corpusString = fs.readFileSync('./test_corpus.txt').toString();
  corpusString = corpusString.replace(/\r/g, '');
  var corpus = corpusString.split("\n");
  for (var i=0; i < corpus.length; i++) {
    if (corpus[i].length === 0) {
      continue;
    }
    var val = hulkify(corpus[i], 140);
    console.log(val);
  }
}

if (require.main === module) {
  main();
}
